engine/resources/hetzner.go

```diff
@@ -0,0 +1,716 @@
+	}
+
+	// verify state specifiers (only allow steady states)
+	switch obj.State {
+	case StateRunning, StateOff, StateAbsent:
+		// continue, state is valid.
+	default:
+		// no other specifiers supported
+		return fmt.Errorf("invalid State")
+	}
+
+	// validate rescue mode options
+	switch obj.ServerRescueMode {
+	case ServerRescueTypeLinux32, ServerRescueTypeLinux64, ServerRescueTypeFreeBSD64:
+		// valid options for rescue mode image.
+	case ServerRescueDisabled:
+		// valid option to disable rescue mode.
+	default:
+		return fmt.Errorf("invalid ServerRescueMode")
+	}
+
+	return nil
+}
+
+func (obj *HetznerRes) Init(init *engine.Init) error {
+
+	// save init struct
+	obj.init = init
+
+	// initialize hcloud client
+	content, err := readFileLines(obj.APITokenPath)
+	token := content[0] // expecting only one!
+	if err != nil {
+		return errwrap.Wrapf(err, "error reading token in obj.APITokenPath")
+	}
+	obj.client = hcloud.NewClient(hcloud.WithToken(token))
+
+	// convert ServerRescueMode to rescueEnable flag
+	obj.rescueEnable = (obj.ServerRescueMode != ServerRescueDisabled)
+
+	return nil
+}
+
+// TODO: clear the root password info if rescue mode was enabled?
+func (obj *HetznerRes) Close() error {
+	obj.client = nil
+	obj.server = nil
+	return nil
+}
+
+func (obj *HetznerRes) Watch() error {
+	return fmt.Errorf("invalid Watch() call: Hetzner has no event API and requires polling")
+}
+
+func (obj *HetznerRes) CheckApply(apply bool) (bool, error) {
+	var checkOk = true
+	ctx := context.TODO()
+	// request up-to-date server info from the API
+	if err := obj.getServerUpdate(ctx); err != nil {
+		return false, err
+	}
+	// Check if the server instance is in the correct state
+	if c, err := obj.checkApplyServerState(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "")
+	} else if !c {
+		checkOk = false
+	}
+	// skip property checks if not in the intended state
+	if !apply && !checkOk {
+		return false, nil
+	}
+	// Check if the active server requires updates
+	if obj.State == StateRunning {
+		// changes in cpu, location and/or image require a server rebuild!
+		if c, err := obj.checkApplyServerRebuild(ctx, apply); err != nil {
+			return false, errwrap.Wrapf(err, "checkApplyServerRebuild failed")
+		} else if !c {
+			checkOk = false
+		}
+		// checkapply rescue mode (update live server)
+		if c, err := obj.checkApplyRescueMode(ctx, apply); err != nil {
+			return false, errwrap.Wrapf(err, "checkApplyServerRebuild failed")
+		} else if !c {
+			checkOk = false
+		}
+		// TODO: checkapply SSH keys
+	}
+	return checkOk, nil
+}
+
+// checkApplyServerState powers the server on/off, creates a new instance, or
+// deletes the existing one as needed, to ensure the correct state.
+func (obj *HetznerRes) checkApplyServerState(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyServerState(apply: %t)", apply)
+	}
+	var checkOk = true
+	if obj.server == nil {
```

Comment:
You might be able to unindent a bunch here and improve readability but I didn't check for sure:
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow